Section 7: Subsystems, and quick ref guide

binex-dsk opened the pull request:
This will complete section 7: subsystems, as well as add a quick reference guide for naming and code conventions, and simple, repetitive, menial tasks.

To do:
- [ ] Finish section 7
  * [x] Creating a subsystem
  * [x] Setting up an instance and getInstance function
  * [ ] Creating variables within the subsystem
  * [ ] What to do with the constructor
  * [ ] Creating methods and getter/setter functions
  * [ ] Constants within a subsystem
  * [x] Where to divide subsystems, and what to include in a single subsystem
- [ ] Finish naming conventions
  * [ ] Review
- [ ] Add code conventions
  * [ ] 4-space tabs
  * [ ] Hierarchy & file structure
  * [ ] etc.
- [ ] Simple task guide
  * [ ] Creating a subsystem/command
  * [ ] Setting up a full command-based robot project (maybe use a template for BeakLib?)
  * [ ] Setting up BeakLib?
  * [ ] Creating a basic project
  * [ ] Deploying code to the robot
  * [ ] Checking controllers
  * [ ] "pre-match" checkup: controllers, etc.

crueter committed:
add quick reference guide and start subsystems

Signed-off-by: swirl <swurl@swurl.xyz>

crueter committed:
<ref.rst> fix indentation

Signed-off-by: swirl <swurl@swurl.xyz>

ghost commented:
Thinking the quick reference guide could have an example of a subsystem made with all improper naming (with comments indicating what's wrong) and another subsystem with everything right (with comments indicating what's right)

crueter committed:
<all> Fix all inline code blocks

Signed-off-by: swirl <swurl@swurl.xyz>

crueter committed:
<sect7.rst> Add "What Are Subsystems?" Section

Signed-off-by: swirl <swurl@swurl.xyz>

crueter committed:
Added Subsystem Setup

<sect7.rst> Added "Creating a Subsystem", "Setting up a Basic Subsystem", and
"Setting up a Subsystem Instance".

<ref.rst> Added "Commonly Used Code" and "Instances".

<sect2.rst> Fixed motor CAN ID.

Signed-off-by: swirl <swurl@swurl.xyz>

crueter committed:
<sect7.rst> Fix some title capitalization

Signed-off-by: swirl <swurl@swurl.xyz>

crueter committed:
<sect7.rst> Fix cross-reference to quick reference guide

Signed-off-by: swirl <swurl@swurl.xyz>

crueter committed:
<sect7.rst> Fix cross-reference again

Signed-off-by: swirl <swurl@swurl.xyz>

crueter committed:
<sect7.rst> BRO

Signed-off-by: swirl <swurl@swurl.xyz>

gabeseq reviewed (approved):
subsystems

crueter committed:
<sect7.rst> Add "Implementing in RobotContainer" section.

Signed-off-by: swirl <swurl@swurl.xyz>

binex-dsk commented on docs/source/ref.rst:
```diff
@@ -0,0 +1,78 @@
+* Public variables of a subsystem/class: ``VariableName`` (title case, no prefix)
+   * Example: ``public double TargetSpeed``. This should generally never be used.
+* "Temporary" variables, within a function or algorithm: ``variableName`` (camel case, no prefix)
+   * Example: ``double targetSpeed``. This should be used within a function or algorithm to store a variable temporarily. For example, if a function takes in a distance in meters, and you need to change it to centimeters for it to be useful, then you can create a temporary variable; i.e. ``double distanceFeet = Units.metersToFeet(distanceMeters)``. You don't need to specify private or public.
```
Oops I meant feet

binex-dsk commented on docs/source/ref.rst:
```diff
@@ -0,0 +1,78 @@
+* "Temporary" variables, within a function or algorithm: ``variableName`` (camel case, no prefix)
+   * Example: ``double targetSpeed``. This should be used within a function or algorithm to store a variable temporarily. For example, if a function takes in a distance in meters, and you need to change it to centimeters for it to be useful, then you can create a temporary variable; i.e. ``double distanceFeet = Units.metersToFeet(distanceMeters)``. You don't need to specify private or public.
+* Constant variables: ``VARIABLE_NAME`` (capital snake case, no prefix)
+   * Example: ``private final double DEFAULT_TARGET_SPEED``. Constants are values that should NEVER change during usage of the robot, and only can be changed by redeploying code. This is useful for defaults, PID values, and similar unchanging values. Ensure that it's ``final``, as that will ensure the value can not be reassigned. If the variable is in a subsystem, then make it private; however, i ...
```
Add incorrect/correct subsystem examples

binex-dsk commented on docs/source/sect7.rst:
```diff
@@ -1,4 +1,182 @@
+
+If that made no sense to you, good. If it did make sense, even better. Here are some examples of what a subsystem might look like.
+
+.. note:: 
```
yes

binex-dsk commented on docs/source/sect7.rst:
```diff
@@ -1,4 +1,182 @@
+.. note:: 
+	TODO: add pictures?
+
+**Drivetrain**: A drivetrain, well, drives a robot! This is *the* universal constant for every competitive FRC robot in existence. It doesn't matter the game or how good the robot is, it has a drivetrain. A drivetrain consists of the following:
```
2002 Newton Finals

binex-dsk commented on docs/source/sect7.rst:
```diff
@@ -1,4 +1,182 @@
+  :alt: RobotContainer with subsystem bindings set up
+  :width: 600
+
+When there are no errors left and everything looks good, deploy your code. Try out your newly subsystem-ified buttons and see if they work! If not, ask a mentor or captain for help (and yes, screaming "CARSON!!!!" across the room is perfectly acceptable.)
```
next sections:
- Constants
- member variables
-idk
